### ISSUE TITLE
statsByDimensionsSync generates wrong src on first run

### ISSUE THREAD

**Issue by timonforrer** [11ty/eleventy-img#146]
I'm using sanity studio and their portable text with [a serializer](https://github.com/portabletext/to-html), to turn the portable text into html.

The serializer is a shortcode, that looks like this:

```js
const { toHTML } = require('@portabletext/to-html');
const generateImage = require('./image');

const blocks = {
  types: {
    imageWithCaption: ({ value: { image } }) => {
      const {
        width,
        height
      } = image.asset.metadata.dimensions;
      const imageHTML = generateImage({
        src: image.asset.url,
        alt: image.alt,
        sizes: "(min-width: 732px) 80ch, 100vw",
        width,
        height
      });
      return imageHTML;
    }
  }
}

module.exports = function renderHTML(input) {
  return toHTML(input, { components: blocks });
}
```

And the `generateImage()` function is:

```js
const Image = require('@11ty/eleventy-img');

module.exports = function imageSyncShortcode(props) {

  const {
    src,
    width,
    height,
    alt,
    sizes,
    loading,
    decoding
  } = props;

  const image_attributes = {
    alt,
    sizes,
    loading: loading ?? 'lazy',
    decoding: decoding ?? 'async',
  }

  const options = {
    widths: [320, 400, 700, 1200, 2000],
    formats: ['avif', 'webp', 'jpeg'],
    urlPath: '/images/',
    outputDir: './dist/images/'
  }

  Image(src, options);

  let metadata = Image.statsByDimensionsSync(src, width, height, options);

  return Image.generateHTML(metadata, image_attributes);
}
```

I'm using `Image.statsByDimensionsSync()` so that I don`t have to use async – I tried the async way before, but that didn't work in conjunction with the sanity serializer.

I found that `Image.statsByDimensionsSync()` generates a different `src` on the first run. So when I just build the site, the generated html for the image tries to load a file that doesn't exist. When I use eleventy serve and let eleventy rebuild the site once, it works fine.

I assume that eleventy generates the filename based on a different value on the first run than on subsequent runs.

E.g. on the first run the src is `/images/CWhDmUxMUw-320.jpeg`, on subsequent runs it's `/images/Zzo8VdfzGM-320.jpeg`

**Comment by timonforrer** [comment]
Alright, I was able to lean on sanity's on-demand image processing to generate the HTML for a responsive image in a synchronous fashion:

```js
const client = require('../sanity.js'); // fully configured sanity client, see https://www.sanity.io/docs/js-client
const imageUrlBuilder = require('@sanity/image-url');

const builder = imageUrlBuilder(client);

module.exports = function sanityImage({ source, alt, aspect_ratio, sizes = '100w', loading = 'lazy', decoding = 'async' }) {

  if (alt === undefined) throw new Error('Alt has to be specified');

  // get essential data from asset
  const {
    metadata: {
      dimensions: {
        width: _w,
        aspectRatio: _aspect
      }
    }
  } = source.asset;

  // define widths for output
  let widths = [320, 400, 700, 1200, 2000];

  // filter out widths that are smaller than the actual image
  widths = widths.filter(w => w <= _w);

  // specify output formats and corresponding mimetype
  const formats = {
    webp:'image/webp',
    jpg: 'image/jpeg'
  };

  // generate sources for each format and width
  const sources = Object.keys(formats).map(f => {
    const _sources = widths.map(w => {
      let url, h;
      if (aspect_ratio !== undefined) {
        // custom aspect ratio defined --> crop image
        const [aspect_w, aspect_h] = aspect_ratio.split('x');
        const dec_aspect_ratio = aspect_w / aspect_h;
        h = Math.round(w / dec_aspect_ratio);
        url = builder.image(source).width(w).height(h).format(f).url();
      } else {
        // no custom aspect ratio, just specify width
        h = Math.round(w / _aspect)
        url = builder.image(source).width(w).format(f).url();
      }
      return {
        url,
        w,
        h
      }
    });

    return {
      format: f,
      mimetype: formats[f],
      srcsets: _sources.map(_s => `${_s.url} ${_s.w}w`).join(', '),
      srcs: _sources
    }
  });

  // get last item of `sources` (jpeg) and select first item of `srcs`, which is the smallest size
  const img = sources[sources.length - 1].srcs[0];

  // generate HTML
  const html = `
  <picture>
    ${sources.map(_s => {
      return `
        <source
          type="${_s.mimetype}"
          srcset="${_s.srcsets}"
          sizes="${sizes}"
        />
      `
    }).join('')}
    <img
      alt="${alt}"
      loading="${loading}"
      decoding="${decoding}"
      src="${img.url}"
      width="${img.w}"
      height="${img.h}"
    />
  </picture>`;

  return html;
}
```

**Comment by zeroby0** [comment]
> E.g. on the first run the src is /images/CWhDmUxMUw-320.jpeg, on subsequent runs it's /images/Zzo8VdfzGM-320.jpeg

Is the image remote?

The filename should be generated by this function: https://github.com/11ty/eleventy-img/blob/d9141cbbdfeece5f2d3d36ea0edcc3f8c5779b3e/img.js#L294

It reads the contents of the file, and hashes them with SHA256. Usually.
But if it's a remote image, it hashes the URL, and checks if the image in cache is valid.
https://github.com/11ty/eleventy-img/blob/d9141cbbdfeece5f2d3d36ea0edcc3f8c5779b3e/img.js#L314

That's a place where the hash could change.

---

Can you see if this bug is reproduceable? If you could make a repository that reproduces this bug, I'll debug it and possibly draft a pull request to fix it.

**Comment by timonforrer** [comment]
@zeroby0 Yes, the image is remote.

The bug should be reproducible – I'll share some code with in the next few days, if you wan't to debug it 🙌🏼

**Comment by timonforrer** [comment]
@zeroby0 I finally took some time trying to reproduce the bug. Appearently I was using v1.0.0 of the img package. It works with v2.0.0.

https://github.com/timonforrer/11ty-statsByDimensionsSync/blob/main/package.json

I'll close this issue

**Comment by timonforrer** [comment]
Hi again, @zeroby0

I was a bit hasty.. Bumping the version to v2.0.0 on the repository, that was originally affected by the bug, didn't resolve the issue.

I added a layout and generate the pages using the pagination feature on the reproduction repo. Now the behaviour is as originally described – the img src is wrong on the first build.

https://github.com/timonforrer/11ty-statsByDimensionsSync/

I would be very glad, if you could have a look at it. Thank you!

**Comment by zeroby0** [comment]
Heyo @timonforrer 

Don't worry, it happens :)

Your repo was very useful in tracing the bug, and it's from this line:
https://github.com/11ty/eleventy-img/blob/d9141cbbdfeece5f2d3d36ea0edcc3f8c5779b3e/img.js#L314

When running for the first time, `eleventy-fetch` fetches your url and caches it, but `this.assetCache.isCacheValid()` returns false. So the hash updated with `ValidCache:false`

In the second run, the image is fetched from the cache and `this.assetCache.isCacheValid()` returns true, and the hash is updated with `ValidCache:true`.

This makes the hash change and returns the wrong src. 

The fix is to make https://github.com/11ty/eleventy-fetch .assetCache.isCacheValid() return `true` on the first run and as long as the cache is valid. I'm not familiar with `eleventy-fetch`'s code, but I'll try to write a patch for it and submit a PR there.

**Comment by timonforrer** [comment]
@zeroby0 thank you for looking into it and creating the PR!

**Comment by zeroby0** [comment]
Thanks for reporting the bug @timonforrer  😄

**Comment by Olivier-Rasquin** [comment]
Hello @zeroby0 , I have the exact same issue described by @timonforrer . 
Any idea when/if your PR will be validated soon?

Thanks

**Comment by zeroby0** [comment]
@Olivier-Rasquin I don't know either, I'm eager for it to be reviewed and merged!

I should find a way to gently pester @zachleat about it XD

**Comment by Olivier-Rasquin** [comment]
Thanks @zeroby0 ! We've just decided that we will wait for the official "update", and in the meantime we are triggering two successive builds.

Let's hope it can be reviewed fairly soon!

**Comment by zeroby0** [comment]
I hope so too! Good luck!

**Comment by zeroby0** [comment]
Hey @zachleat !

Hope you had a great holiday! 

When you have the time, can you please take a look at this issue and the fix for it in #146? There's a race condition when fetching remote images causing hash to change.

**Comment by zachleat** [comment]
Howdy y’all—sorry for the super late reply here but this is still a bug in Eleventy Image and I’m evaluating the best solution here.

I can verify that the `useCacheValidityInHash: false` option (introduced in Eleventy Image v1.0.0 but not currently on the docs for this 😭, building now) is a viable workaround for this issue for now to have consistent file names in the test case repo provided at: https://github.com/timonforrer/11ty-statsByDimensionsSync/

All that being said, we’re going to remove the `useCacheValidityInHash` option in v6.0.0—cache validity will no longer play a role in how hashes are generated for caches moving forward. Instead, we’re going to add cache validity as a signal to whether or not we write images to disk.

Regardless, this requires a new Fetch version and I’ve retested the test case attached here to make sure it works.

This fix will ship with v6.0.0-beta.1.
